net:rpl: Null pointer dereferences

satyacube opened the pull request:
Check for NULL IPv6 addr values from net_if_ipv6_get_ll() in rpl.c
If NULL, print a debug statement
Coverity-CID: 188169

Fixes #10094.

Signed-off-by: Satya Bhattacharya <satyacube@gmail.com>

jukkar commented on subsys/net/ip/rpl.c:
```diff
@@ -649,13 +649,20 @@ static void dio_timer(struct k_work *work)
+			if (addr) {
+				net_rpl_dio_send(rpl_default_iface, instance, addr,
+					         NULL);
+				#if defined(CONFIG_NET_STATISTICS_RPL)
```
We typically do not use indented pre-processor code, so #if should start from first column.

jukkar replied:
Please start the #if from the column 0

jukkar reviewed (changes requested):
Looks ok, minor tweak needed. See also https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/24243/5/tests about the errors the sanitychecker is reporting.

pfalcon reviewed (changes requested):
Commit message doesn't correspond to the change made. The commit message says:

> Check for NULL IPv6 addr values from net_if_ipv6_get_ll() in rpl.c
If NULL, print a debug statement

But that doesn't describe what you're doing fully. Because besides printing debug statement, you actually schedule a delayed work. IMHO, the commit message should explain why you're doing that (I for one can't assess if that's a right change or not).

pfalcon commented:
recheck

pfalcon commented:
@satyacube:

Thanks for the updates!

> Do you guys know how to trigger the Shippable sanity runs?

On each new push it should trigger itself, and if something goes wrong, you can manually trigger it by posting comment of just "recheck" (like I did above).

pfalcon reviewed (dismissed):
Thanks for the updates! Let me approve from my side, but I hope people more knowledge in RPL will eyeball it too.

jfischer-phytec-iot commented on subsys/net/ip/rpl.c:
```diff
@@ -650,12 +650,20 @@ static void dio_timer(struct k_work *work)
+				#if defined(CONFIG_NET_STATISTICS_RPL)
+				instance->dio_send_pkt++;
+				#endif
+			} else {
```
The case (addr == NULL) will only occur if, for example, an interface(802.15.4) do not exist or something went wrong when initializing the driver. I do not know how much sense it makes to resubmit here. See also the else branch below (L667). @rveerama1 Can you comment on that?

rveerama1 replied:
I agree with @jfischer-phytec-iot. This doesn't make sense to resubmit after timeout. You end up with same addr is NULL. It's better to through NET_ERR message and find the root cause of addr being NULL. Do you have any test case or sample app with this scenario other than coverity issue?

jfischer-phytec-iot replied:
I observed it when I did not plug in FRDM-CR20A properly while testing rpl_border_router. To provoke it, it is enough to try rpl_border_router or rpl_node on FRDM-K64F without a FRDM-CR20A.

satyacube committed:
net: rpl: Null pointer dereferences

Check for NULL IPv6 addr values from net_if_ipv6_get_ll() in rpl.c
If NULL, print out an error statement stating that no proper IPv6
address was found
Coverity-CID: 188169

Fixes #10094.

Signed-off-by: Satya Bhattacharya <satyacube@gmail.com>

jukkar commented on subsys/net/ip/rpl.c:
```diff
@@ -650,12 +650,18 @@ static void dio_timer(struct k_work *work)
+			} else {
+				NET_ERR("Cannot send DIO, IPv6 link local "
+					"address is NULL");
+			return;
```
Indentation wrong here

jukkar commented:
BTW, there is no need to constantly close/re-open this PR. If you have a new version, just force push it over the old thread. The PR will be closed automatically when it is merged to upstream.

jukkar commented on subsys/net/ip/rpl.c:
```diff
@@ -650,12 +650,18 @@ static void dio_timer(struct k_work *work)
+						 addr, NULL);
 #if defined(CONFIG_NET_STATISTICS_RPL)
-			instance->dio_send_pkt++;
+	instance->dio_send_pkt++;
```
The indentation of this line is wrong, it should be placed in same column as the net_rpl_dio_send() above.

satyacube replied:
Ah! I missed that one. Thanks!
